README.md
![Bacteriomatics avatar white](https://github.com/Volnikan/Bacteriomatics/assets/95049684/79fdde1c-47b0-4655-a9a1-764748f00988)

# Bacteriomatics

Have you ever dreamed about creating your own species of creatures? If yes, this is what you're going to do in this game!

## Game objective

You are a talented and ambitious scientist, working at a scientifical research center, who wants to become its new director. Build your career from a laboratory assistent to the director of the whole scientific organization!

## General game mechanics
### Accept contracts

Accept commercial, government or scientific contracts, complete them and get money and rating. This will help you to grow professionally.

### Cause mutations

The main idea of the game is to cause specific mutations in the genetic code of bacterias. Your goal is to make bacterias produce certain types of substances, depending on the conditions of your current contract.

### Get the result

When you achieve creating a new species, which meets all the requirements, close the contract and get the reward!

### Upgrade your equipment and buy new stuff

Now you can upgrade your equipment and buy some new mutagens for making your work more effective.


CHANGELOG.md
# Changelog


## [0.6.0-alpha](https://github.com/Volnikan/Bacteriomatics/compare/v0.5.1-alpha...v0.6.0-alpha) (2023-07-21)


### Features

* add buttons to control zoom values ([37b7623](https://github.com/Volnikan/Bacteriomatics/commit/37b7623040a363c9af186e86f42c61306e5359f9))
* add displayed values to volume sliders ([5b542d4](https://github.com/Volnikan/Bacteriomatics/commit/5b542d4e3e0b6e9bd11add095058572b1f058555)), closes [#23](https://github.com/Volnikan/Bacteriomatics/issues/23)
* add menu button ([f1118fe](https://github.com/Volnikan/Bacteriomatics/commit/f1118fedfad191607b3b48437c072c52c05e9183)), closes [#14](https://github.com/Volnikan/Bacteriomatics/issues/14)
* create release-please.yml ([5fecd53](https://github.com/Volnikan/Bacteriomatics/commit/5fecd53446f98dff34f64b5b4ea7e799e1a5164b))
* hide status bar ([102414e](https://github.com/Volnikan/Bacteriomatics/commit/102414e926785f3d099af6b0db1a9ff142c52ee0))
* save volume settings into device's memory ([365c6e6](https://github.com/Volnikan/Bacteriomatics/commit/365c6e6bb46bec902f7cb232697c3bbfab3161a1)), closes [#25](https://github.com/Volnikan/Bacteriomatics/issues/25)


### Bug Fixes

* center the position of symbols on buttons ([590b491](https://github.com/Volnikan/Bacteriomatics/commit/590b491fc48225f1588319393b8fabdafe1f45ba)), closes [#35](https://github.com/Volnikan/Bacteriomatics/issues/35)
* fix app icons view for Android 8 and later ([6ce57c2](https://github.com/Volnikan/Bacteriomatics/commit/6ce57c2c46084e45683cecbcaba759b4f35ddbaf)), closes [#33](https://github.com/Volnikan/Bacteriomatics/issues/33)
* open the Windows version in fullscreen ([43b49af](https://github.com/Volnikan/Bacteriomatics/commit/43b49af240e0909005be95cb70859ca3a6af6f64)), closes [#32](https://github.com/Volnikan/Bacteriomatics/issues/32)
* **settings:** fix runtime error ([b73605f](https://github.com/Volnikan/Bacteriomatics/commit/b73605ff775086b9ca610323f42803505c8f4ded))


### Documentation Changes

* create CONTRIBUTING.md ([4decc57](https://github.com/Volnikan/Bacteriomatics/commit/4decc573dea16beed169f07c01d0e809ef6627b6)), closes [#36](https://github.com/Volnikan/Bacteriomatics/issues/36)
* update README.md ([faf3d01](https://github.com/Volnikan/Bacteriomatics/commit/faf3d01f8784430f6c6f47f89332e986377ef8a0)), closes [#37](https://github.com/Volnikan/Bacteriomatics/issues/37)


### Miscellaneous

* add configuration .xml files for the icons ([bfa2d63](https://github.com/Volnikan/Bacteriomatics/commit/bfa2d6303ad4332d95093ad2fbea92f4370d4e1b)), closes [#13](https://github.com/Volnikan/Bacteriomatics/issues/13)
* add custom icons for Android 7 and earlier ([77512ce](https://github.com/Volnikan/Bacteriomatics/commit/77512ce01f55eae2590eca5326905e80cbc3a55d))
* add custom icons for Android 8 and later ([3044690](https://github.com/Volnikan/Bacteriomatics/commit/3044690fec9d47f29dbc58447997e992ead22aaf))
* add menu button image sheet ([db26046](https://github.com/Volnikan/Bacteriomatics/commit/db26046c08e5147bb94209ded32ccaeeed92509d))
* add new buttons sound ([4f1395b](https://github.com/Volnikan/Bacteriomatics/commit/4f1395b617143daf0b2f431d97141b0db082ba09))
* delete old button sound ([a90a896](https://github.com/Volnikan/Bacteriomatics/commit/a90a8964ab7e04669117e6de062ef7ed594dbe3b))
* delete test CHANGLELOG.md ([4057918](https://github.com/Volnikan/Bacteriomatics/commit/4057918c2ad91ae1eaf6268c3844767747ef7c0b))
* delete unused stepper image sheet ([fd869e3](https://github.com/Volnikan/Bacteriomatics/commit/fd869e3e95baf7d8ec72979ff13d93d22ef042d9))
* **main:** release 0.5.0-alpha ([39cc16d](https://github.com/Volnikan/Bacteriomatics/commit/39cc16df659925d0f31730f4be01a9ed2d4e6e1a))
* **main:** release 0.5.0-alpha ([cce2ba0](https://github.com/Volnikan/Bacteriomatics/commit/cce2ba04f93d71d2eab67a4a0d28145788c7e836))
* **main:** release 0.5.1-alpha ([ded07fb](https://github.com/Volnikan/Bacteriomatics/commit/ded07fb955d133db2a4d15ed33c3569351a1fe73))
* make the buttons sound on tap ([c21b0fb](https://github.com/Volnikan/Bacteriomatics/commit/c21b0fbae247632fa0f3c1f09a4bef8e0b7c59f6)), closes [#27](https://github.com/Volnikan/Bacteriomatics/issues/27)
* relocate scene .lua files into Scenes/ ([83704c2](https://github.com/Volnikan/Bacteriomatics/commit/83704c2547eb81385d196d1d9289f64d54b878e8)), closes [#38](https://github.com/Volnikan/Bacteriomatics/issues/38)

## [0.5.0-alpha](https://github.com/Volnikan/Bacteriomatics/compare/v0.4.1-alpha...v0.5.0-alpha) (2023-07-18)


### Features

* add "play" and "settings" butons ([64994e6](https://github.com/Volnikan/Bacteriomatics/pull/7/commits/64994e67497de724c0a55c245b24b4faae706a4a))
* add buttons to control zoom values ([37b7623](https://github.com/Volnikan/Bacteriomatics/commit/37b7623040a363c9af186e86f42c61306e5359f9))
* add displayed values to volume sliders ([5b542d4](https://github.com/Volnikan/Bacteriomatics/commit/5b542d4e3e0b6e9bd11add095058572b1f058555)), closes [#23](https://github.com/Volnikan/Bacteriomatics/issues/23)
* add menu button ([f1118fe](https://github.com/Volnikan/Bacteriomatics/commit/f1118fedfad191607b3b48437c072c52c05e9183)), closes [#14](https://github.com/Volnikan/Bacteriomatics/issues/14)
* create menu scene ([34e0851](https://github.com/Volnikan/Bacteriomatics/pull/7/commits/34e0851eec147a9214fe90a624696c67b9830a9a))
* hide status bar ([102414e](https://github.com/Volnikan/Bacteriomatics/commit/102414e926785f3d099af6b0db1a9ff142c52ee0))
* save volume settings into device's memory ([365c6e6](https://github.com/Volnikan/Bacteriomatics/commit/365c6e6bb46bec902f7cb232697c3bbfab3161a1)), closes [#25](https://github.com/Volnikan/Bacteriomatics/issues/25)

### Bug Fixes

* **settings:** fix runtime error ([b73605f](https://github.com/Volnikan/Bacteriomatics/commit/b73605ff775086b9ca610323f42803505c8f4ded))

### Miscellaneous


Scenes/settings.lua
local composer = require("composer")
local widget = require("widget")
local json = require("json")
local scene = composer.newScene()

-- Setting UI color palette
local uiColorDark = {0.13, 0.13, 0.13}
local uiColorLight = {0.94, 0.94, 0.94}
local uiColorGreenDark = {0.01, 0.15, 0.07}
local uiColorGreenMedium = {0.02, 0.46, 0.22}
local uiColorGreenLight = {0.15, 0.97, 0.52}

-- Creating variables
local musicVolume = 50
local soundVolume = 50
local musicVolumeText
local soundVolumeText
local settingsParametersPath = system.pathForFile("Settings.json", system.DocumentsDirectory)

-- Sounds variables
local buttonSound

-- Settings table
local settingsTable = {
	["language"] = "",
	["music"] = musicVolume,
	["sound"] = soundVolume
}

-- Slider imagesheet options
local sliderOptions = {
	frames = {
		{x = 4, y = 76, width = 32, height = 32},
		{x = 44, y = 76, width = 32, height = 32},
		{x = 76, y = 4, width = 32, height = 32},
		{x = 76, y = 36, width = 32, height = 32},
		{x = 4, y = 4, width = 64, height = 64}
	},
	sheetContentWidth = 112,
	sheetContentHeight = 112
}

-- Reading settings function
local function readSettings()
	
	local settingsFile = io.open(settingsParametersPath, "r")
	
	if settingsFile then
		local contents = settingsFile:read("*a")
		settingsTable = json.decode(contents)
		musicVolume = settingsTable.music
		soundVolume = settingsTable.sound
		io.close(settingsFile)
	end
	
	settingsFile = nil
	
end

-- Writitng settings function
local function writeSettings()
	
	settingsTable.music = musicVolume
	settingsTable.sound = soundVolume
	
	local settingsFile = io.open(settingsParametersPath, "w")
	settingsFile:write(json.encode(settingsTable))
	io.close(settingsFile)
	settingsFile = nil
	
end

-- Music volume slider function
local function onMusicSliderMove(event)

	musicVolume = event.value
	musicVolumeText.text = "Music volume: " .. musicVolume
	
	if(event.phase == "ended") then
		writeSettings()
	end
end

-- Sound volume slider function
local function onSoundSliderMove(event)

	soundVolume = event.value
	soundVolumeText.text = "Sound volume: " .. soundVolume
	
	if(event.phase == "ended") then
		writeSettings()
	end
end

-- Back button function
local function onBackButton(event)
	
	if(event.keyName == "back") then
		composer.gotoScene("Scenes.menu")
		return true
	end
	
	if(event.phase == "began") then
		audio.play(buttonSound)
	end
	
	if(event.phase == "ended") then
		composer.gotoScene("Scenes.menu")
	end

end

-- SCENE:CREATE
function scene:create(event)
	
	-- Creating and adjusting groups
	local sceneGroup = self.view
	local bgGroup = display.newGroup()
	local uiGroup = display.newGroup()
	local textGroup = display.newGroup()
	sceneGroup:insert(bgGroup)
	sceneGroup:insert(uiGroup)
	sceneGroup:insert(textGroup)
	
	-- Creating background
	local background = display.newRect(bgGroup, display.contentCenterX, display.contentCenterY, display.contentWidth, display.contentHeight)
	background:setFillColor(unpack(uiColorDark))
	
	-- Displaying settings
	
	readSettings()
	
	-- Language settings
	local languageText = display.newText(textGroup, "Language: ", 120, 120, native.systemFontBold, 60)
	languageText.anchorX = 0
	languageText.anchorY = 0.5
	languageText:setFillColor(unpack(uiColorLight))
	
	-- Music volume settings
	musicVolumeText = display.newText(textGroup, "Music volume: " .. musicVolume, 120, 320, native.systemFontBold, 60)
	musicVolumeText.anchorX = 0
	musicVolumeText.anchorY = 0.5
	languageText:setFillColor(unpack(uiColorLight))
	
	-- Music volume slider
	local sliderSheet = graphics.newImageSheet("Images/Sliders/SliderImageSheet.png", sliderOptions)
	
	local musicSlider = widget.newSlider(
		{
			sheet = sliderSheet,
			leftFrame = 1,
			middleFrame = 2,
			rightFrame = 3,
			fillFrame = 4,
			frameWidth = 48,
			frameHeight = 48,
			handleFrame = 5,
			handleWidth = 96,
			handleHeight = 96,
			x = display.contentWidth - 600,
			y = 320,
			orientation = "horizontal",
			width = 800,
			value = musicVolume,
			listener = onMusicSliderMove
		}
	)
	
	uiGroup:insert(musicSlider)
	musicSlider.anchorX = 1
	musicSlider.anchorY = 0.5
	
	-- Sound volume settings
	soundVolumeText = display.newText(textGroup, "Sound volume: " .. soundVolume, 120, 520, native.systemFontBold, 60)
	soundVolumeText.anchorX = 0
	soundVolumeText.anchorY = 0.5
	soundVolumeText:setFillColor(unpack(uiColorLight))
	
	local soundSlider = widget.newSlider(
		{
			sheet = sliderSheet,
			leftFrame = 1,
			middleFrame = 2,
			rightFrame = 3,
			fillFrame = 4,
			frameWidth = 48,
			frameHeight = 48,
			handleFrame = 5,
			handleWidth = 96,
			handleHeight = 96,
			x = display.contentWidth - 600,
			y = 520,
			orientation = "horizontal",
			width = 800,
			value = soundVolume,
			listener = onSoundSliderMove
		}
	)
	
	uiGroup:insert(soundSlider)
	soundSlider.anchorX = 1
	soundSlider.anchorY = 0.5
	
	-- Creating buttons
	
	-- Back button
	local buttonBack = widget.newButton({
		label = "back_button",
		onEvent = onBackButton,
		shape = "roundedRect",
		width = 300,
		height = 120,
		cornerRadius = 60,
		fillColor = {default = uiColorGreenMedium, over = uiColorGreenLight},
		strokeColor = {default = uiColorGreenLight, over = uiColorGreenMedium},
		strokeWidth = 16,
		labelColor = {default = uiColorLight, over = uiColorDark},
		fontSize = 60
	})
	
	uiGroup:insert(buttonBack)
	buttonBack.anchorX = 0
	buttonBack.anchorY = 0.5
	buttonBack.x = 120
	buttonBack.y = display.contentHeight - 120
	buttonBack:setLabel("Back")
	
	-- Initializing sounds
	buttonSound = audio.loadSound("Sounds/Button_sound1.wav")
	
end

-- SCENE:SHOW
function scene:show(event)
	local sceneGroup = self.view
	local phase = event.phase
	if(event.phase == "will") then
		
	elseif(event.phase == "did") then
		
		Runtime:addEventListener("key", onBackButton)
		
	end
end

-- SCENE:HIDE
function scene:hide(event)
	local sceneGroup = self.view
	local phase = event.phase
	if(event.phase == "will") then
		
	elseif(event.phase == "did") then
		
		Runtime:removeEventListener("key", onBackButton)
		
	end
end

-- SCENE:DESTROY
function scene:destroy(event)
	
	local sceneGroup = self.view
	audio.dispose(buttonSound)
	
end

-- Adding scene event listeners
scene:addEventListener("create", scene)
scene:addEventListener("show", scene)
scene:addEventListener("hide", scene)
scene:addEventListener("destroy", scene)

return scene